File: win_app/src/QueenMama.Core/Services/License/LicenseManager.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using CommunityToolkit.Mvvm.ComponentModel;
using Microsoft.Extensions.Logging;
using QueenMama.Core.Interfaces;
using QueenMama.Core.Models;
using QueenMama.Core.Services.Auth;

namespace QueenMama.Core.Services.License;

/// <summary>
/// Manages license validation and feature gating
/// </summary>
public partial class LicenseManager : ObservableObject, ILicenseManager
{
    private readonly AuthApiClient _api;
    private readonly IAuthenticationManager _authManager;
    private readonly ILogger<LicenseManager> _logger;

    private const string CacheKey = "cached_license";
    private const string CacheExpiryKey = "cached_license_expiry";
    private const string UsageResetKey = "usage_reset_date";
    private const int GracePeriodDays = 7;
    private const int RevalidationIntervalMinutes = 60;

    private Timer? _revalidationTimer;
    private readonly string _settingsPath;

    public event Action<Models.License?>? OnLicenseChanged;

    [ObservableProperty]
    private Models.License _currentLicense = Models.License.Free;

    [ObservableProperty]
    private bool _isValidating;

    [ObservableProperty]
    private bool _isOffline;

    [ObservableProperty]
    private DateTime? _lastValidatedAt;

    [ObservableProperty]
    private int _smartModeUsedToday;

    [ObservableProperty]
    private int _aiRequestsToday;

    public LicenseTier CurrentTier => CurrentLicense.Tier;

    /// <summary>
    /// License secret for HMAC signature verification
    /// </summary>
    private string LicenseSecret
    {
        get
        {
            var envSecret = Environment.GetEnvironmentVariable("LICENSE_SECRET");
            return envSecret ?? BundledLicenseSecret;
        }
    }

    /// <summary>
    /// Bundled license secret (obfuscated)
    /// </summary>
    private static readonly string BundledLicenseSecret = string.Join("",
        new[] { "FpOU+px9", "sASk0/+e", "Zu5uDQBP", "4r0rhZ2g", "2h5u5iqf", "0Uw=" });

    public LicenseManager(
        IAuthApiClient api,
        IAuthenticationManager authManager,
        ILogger<LicenseManager> logger)
    {
        _api = (AuthApiClient)api;
        _authManager = authManager;
        _logger = logger;

        _settingsPath = Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
            "QueenMama", "settings");
        Directory.CreateDirectory(_settingsPath);

        // Load cached license
        LoadCachedLicense();

        // Reset daily usage if needed
        ResetDailyUsageIfNeeded();

        // Subscribe to auth changes
        _authManager.OnAuthStateChanged += OnAuthStateChangedHandler;

        // Start periodic revalidation
        StartPeriodicRevalidation();
    }

    partial void OnCurrentLicenseChanged(Models.License value)
    {
        OnLicenseChanged?.Invoke(value);
    }

    private void OnAuthStateChangedHandler(AuthState state)
    {
        Task.Run(async () =>
        {
            if (state == AuthState.Authenticated)
            {
                await RevalidateAsync();
            }
            else if (state == AuthState.SignedOut)
            {
                CurrentLicense = Models.License.Free;
                ResetUsage();
            }
        });
    }

    private void StartPeriodicRevalidation()
    {
        _revalidationTimer?.Dispose();
        _revalidationTimer = new Timer(
            async _ => await RevalidateIfNeededAsync(),
            null,
            TimeSpan.FromMinutes(RevalidationIntervalMinutes),
            TimeSpan.FromMinutes(RevalidationIntervalMinutes));
    }

    private async Task RevalidateIfNeededAsync()
    {
        if (!_authManager.IsAuthenticated)
            return;

        // Skip if validated recently (within last 5 minutes)
        if (LastValidatedAt.HasValue &&
            DateTime.UtcNow.Subtract(LastValidatedAt.Value).TotalMinutes < 5)
        {
            return;
        }

        await RevalidateAsync();
    }

    public bool HasFeature(Feature feature)
    {
        return CurrentLicense.HasFeature(feature);
    }

    public bool IsFeatureEnabled(Feature feature)
    {
        var access = CheckFeatureAccess(feature);
        return access.IsAllowed;
    }

    public FeatureAccess CheckFeatureAccess(Feature feature)
    {
        // Block all features for unauthenticated users
        if (!_authManager.IsAuthenticated)
        {
            if (feature == Feature.SessionHistory)
                return FeatureAccess.RequiresAuth;
            return FeatureAccess.Blocked;
        }

        var currentUsage = new UsageStats
        {
            SmartModeUsedToday = SmartModeUsedToday,
            AiRequestsToday = AiRequestsToday
        };

        return CurrentLicense.CheckFeatureAccess(feature, currentUsage);
    }

    public void RecordUsage(Feature feature, string? provider = null)
    {
        string action;
        switch (feature)
        {
            case Feature.SmartMode:
                SmartModeUsedToday++;
                SaveUsage();
                action = "smart_mode";
                break;
            case Feature.AIAssist:
                AiRequestsToday++;
                SaveUsage();
                action = "ai_request";
                break;
            case Feature.SessionStart:
                action = "session_start";
                break;
            case Feature.AutoAnswer:
                action = "auto_answer";
                break;
            default:
                return;
        }

        // Record usage to server asynchronously
        _ = Task.Run(async () =>
        {
            try
            {
                await _api.RecordUsageAsync(action, provider);
            }
            catch (Exception ex)
            {
                _logger.LogWarning(ex, "Failed to record usage to server");
            }
        });
    }

    public async Task<Models.License?> ValidateLicenseAsync()
    {
        await RevalidateAsync();
        return CurrentLicense;
    }

    public async Task<Models.License?> FetchLicenseAsync()
    {
        if (!_authManager.IsAuthenticated)
        {
            return Models.License.Free;
        }

        try
        {
            var license = await _api.ValidateLicenseAsync(DeviceInfo.Current().DeviceId);

            if (VerifyLicenseSignature(license))
            {
                return license;
            }
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to fetch license");
        }

        return null;
    }

    private async Task RevalidateAsync()
    {
        if (!_authManager.IsAuthenticated)
        {
            CurrentLicense = Models.License.Free;
            return;
        }

        IsValidating = true;
        IsOffline = false;

        try
        {
            var license = await _api.ValidateLicenseAsync(DeviceInfo.Current().DeviceId);

            if (VerifyLicenseSignature(license))
            {
                CurrentLicense = license;
                LastValidatedAt = DateTime.UtcNow;

                // Update usage from server
                if (license.Usage != null)
                {
                    SmartModeUsedToday = license.Usage.SmartModeUsedToday;
                    AiRequestsToday = license.Usage.AiRequestsToday;
                }

                // Cache the license
                CacheLicense(license);
                _logger.LogInformation("License validated: {Plan}", license.Plan);
            }
            else
            {
                _logger.LogWarning("SECURITY: Invalid license signature detected");

                // Fall back to cached license if valid
                var cachedLicense = LoadCachedLicense();
                if (cachedLicense != null && VerifyLicenseSignature(cachedLicense))
                {
                    CurrentLicense = cachedLicense;
                }
                else
                {
                    CurrentLicense = Models.License.Free;
                }
            }
        }
        catch (Exception ex)
        {
            _logger.LogWarning(ex, "License validation failed, using cached");
            IsOffline = true;

            // Use cached license with grace period
            var cachedLicense = LoadCachedLicense();
            if (cachedLicense != null && VerifyLicenseSignature(cachedLicense))
            {
                var expiry = GetCacheExpiry();
                if (cachedLicense.Plan == SubscriptionPlan.Pro && expiry != null && expiry > DateTime.UtcNow)
                {
                    CurrentLicense = cachedLicense;
                }
                else if (cachedLicense.Plan != SubscriptionPlan.Pro)
                {
                    CurrentLicense = cachedLicense;
                }
                else
                {
                    CurrentLicense = Models.License.Free;
                }
            }
        }

        IsValidating = false;
    }

    private bool VerifyLicenseSignature(Models.License license)
    {
        // Skip verification for local free license (no signature)
        if (string.IsNullOrEmpty(license.Signature) && license.Plan == SubscriptionPlan.Free)
        {
            return true;
        }

        var signedPayload = BuildSignaturePayload(license);
        var payloadBytes = Encoding.UTF8.GetBytes(signedPayload);

        using var hmac = new HMACSHA256(Encoding.UTF8.GetBytes(LicenseSecret));
        var hash = hmac.ComputeHash(payloadBytes);
        var computedSignature = Convert.ToHexString(hash).ToLowerInvariant();

        // Constant-time comparison
        return ConstantTimeCompare(computedSignature, license.Signature);
    }

    private static bool ConstantTimeCompare(string a, string b)
    {
        if (a.Length != b.Length)
            return false;

        var result = 0;
        for (var i = 0; i < a.Length; i++)
        {
            result |= a[i] ^ b[i];
        }

        return result == 0;
    }

    private static string BuildSignaturePayload(Models.License license)
    {
        var sb = new StringBuilder();
        sb.Append('{');
        sb.Append($"\"valid\":{license.Valid.ToString().ToLower()},");
        sb.Append($"\"plan\":\"{license.Plan}\",");
        sb.Append($"\"status\":\"{license.Status}\",");
        sb.Append($"\"features\":{BuildFeaturesJson(license.Features)},");
        sb.Append($"\"trial\":{BuildTrialJson(license.Trial)},");
        sb.Append($"\"cacheTTL\":{license.CacheTTL},");
        sb.Append($"\"validatedAt\":\"{license.ValidatedAt}\"");
        sb.Append('}');
        return sb.ToString();
    }

    private static string BuildFeaturesJson(LicenseFeatures features)
    {
        var sb = new StringBuilder();
        sb.Append('{');
        sb.Append($"\"smartModeEnabled\":{features.SmartModeEnabled.ToString().ToLower()},");
        sb.Append($"\"smartModeLimit\":{(features.SmartModeLimit.HasValue ? features.SmartModeLimit.ToString() : "null")},");
        sb.Append($"\"customModesEnabled\":{features.CustomModesEnabled.ToString().ToLower()},");
        sb.Append($"\"exportFormats\":[{string.Join(",", features.ExportFormats.Select(f => $"\"{f}\""))}],");
        sb.Append($"\"autoAnswerEnabled\":{features.AutoAnswerEnabled.ToString().ToLower()},");
        sb.Append($"\"sessionSyncEnabled\":{features.SessionSyncEnabled.ToString().ToLower()},");
        sb.Append($"\"dailyAiRequestLimit\":{(features.DailyAiRequestLimit.HasValue ? features.DailyAiRequestLimit.ToString() : "null")},");
        sb.Append($"\"maxSyncedSessions\":{(features.MaxSyncedSessions.HasValue ? features.MaxSyncedSessions.ToString() : "null")},");
        sb.Append($"\"maxTranscriptSize\":{(features.MaxTranscriptSize.HasValue ? features.MaxTranscriptSize.ToString() : "null")},");
        sb.Append($"\"undetectableEnabled\":{features.UndetectableEnabled.ToString().ToLower()},");
        sb.Append($"\"screenshotEnabled\":{features.ScreenshotEnabled.ToString().ToLower()}");
        sb.Append('}');
        return sb.ToString();
    }

    private static string BuildTrialJson(TrialInfo? trial)
    {
        if (trial == null)
            return "null";

        return $"{{\"isActive\":{trial.IsActive.ToString().ToLower()},\"daysRemaining\":{trial.DaysRemaining},\"endsAt\":\"{trial.EndsAt}\"}}";
    }

    private Models.License? LoadCachedLicense()
    {
        var path = Path.Combine(_settingsPath, CacheKey);
        if (!File.Exists(path))
            return null;

        try
        {
            var json = File.ReadAllText(path);
            var license = JsonSerializer.Deserialize<Models.License>(json);
            if (license != null)
            {
                CurrentLicense = license;
            }
            return license;
        }
        catch (Exception ex)
        {
            _logger.LogWarning(ex, "Failed to load cached license");
            return null;
        }
    }

    private void CacheLicense(Models.License license)
    {
        try
        {
            var json = JsonSerializer.Serialize(license);
            var path = Path.Combine(_settingsPath, CacheKey);
            File.WriteAllText(path, json);

            // Set expiry with grace period for PRO
            var expiryDays = license.Plan == SubscriptionPlan.Pro ? GracePeriodDays : 1;
            var expiry = DateTime.UtcNow.AddDays(expiryDays);
            var expiryPath = Path.Combine(_settingsPath, CacheExpiryKey);
            File.WriteAllText(expiryPath, expiry.ToString("O"));
        }
        catch (Exception ex)
        {
            _logger.LogWarning(ex, "Failed to cache license");
        }
    }

    private DateTime? GetCacheExpiry()
    {
        var path = Path.Combine(_settingsPath, CacheExpiryKey);
        if (!File.Exists(path))
            return null;

        try
        {
            var str = File.ReadAllText(path);
            return DateTime.Parse(str);
        }
        catch
        {
            return null;
        }
    }

    private void ResetDailyUsageIfNeeded()
    {
        var today = DateTime.UtcNow.Date;
        var resetPath = Path.Combine(_settingsPath, UsageResetKey);

        DateTime? lastReset = null;
        if (File.Exists(resetPath))
        {
            try
            {
                lastReset = DateTime.Parse(File.ReadAllText(resetPath));
            }
            catch { }
        }

        if (lastReset == null || lastReset.Value.Date != today)
        {
            ResetUsage();
            File.WriteAllText(resetPath, today.ToString("O"));
        }
        else
        {
            LoadUsage();
        }
    }

    private void ResetUsage()
    {
        SmartModeUsedToday = 0;
        AiRequestsToday = 0;
        SaveUsage();
    }

    private void LoadUsage()
    {
        var smartModePath = Path.Combine(_settingsPath, "smartModeUsedToday");
        var aiRequestsPath = Path.Combine(_settingsPath, "aiRequestsToday");

        if (File.Exists(smartModePath))
        {
            try { SmartModeUsedToday = int.Parse(File.ReadAllText(smartModePath)); } catch { }
        }

        if (File.Exists(aiRequestsPath))
        {
            try { AiRequestsToday = int.Parse(File.ReadAllText(aiRequestsPath)); } catch { }
        }
    }

    private void SaveUsage()
    {
        var smartModePath = Path.Combine(_settingsPath, "smartModeUsedToday");
        var aiRequestsPath = Path.Combine(_settingsPath, "aiRequestsToday");

        File.WriteAllText(smartModePath, SmartModeUsedToday.ToString());
        File.WriteAllText(aiRequestsPath, AiRequestsToday.ToString());
    }
}

File: mac_app/Services/LicenseManager.swift
import Foundation
import Combine
import CryptoKit

/// Manages license validation and feature gating
@MainActor
final class LicenseManager: ObservableObject {
    static let shared = LicenseManager()

    // MARK: - Published State

    @Published private(set) var currentLicense: License = .free
    @Published private(set) var isValidating: Bool = false
    @Published private(set) var isOffline: Bool = false
    @Published private(set) var lastValidatedAt: Date?
    @Published private(set) var lastSignatureError: String?

    // MARK: - Usage Tracking

    @Published private(set) var smartModeUsedToday: Int = 0
    @Published private(set) var aiRequestsToday: Int = 0

    // MARK: - Dependencies

    private let api = AuthAPIClient.shared
    private let authManager = AuthenticationManager.shared
    private let deviceInfo = DeviceInfo.current()

    // MARK: - Cache

    private let cacheKey = "cached_license"
    private let cacheExpiryKey = "cached_license_expiry"
    private let usageResetKey = "usage_reset_date"
    private let gracePeriodDays = 7
    private let revalidationIntervalMinutes = 60 // Revalidate every hour

    private var cancellables = Set<AnyCancellable>()
    private var revalidationTimer: Timer?

    // MARK: - Signature Verification

    /// License secret for HMAC signature verification
    /// SECURITY: The secret is now retrieved ONLY from environment or server
    /// Never bundle secrets in the binary - they can be extracted
    private var licenseSecret: String? {
        // Only retrieve from environment - never use a bundled fallback
        // Set LICENSE_SECRET in your Xcode scheme environment variables
        // or retrieve it securely from the server during initial auth
        ProcessInfo.processInfo.environment["LICENSE_SECRET"]
    }

    private init() {
        // Load cached license
        loadCachedLicense()

        // Reset daily usage if needed
        resetDailyUsageIfNeeded()

        // Subscribe to auth changes
        authManager.$isAuthenticated
            .dropFirst()
            .sink { [weak self] isAuthenticated in
                Task { @MainActor in
                    if isAuthenticated {
                        await self?.revalidate()
                    } else {
                        self?.currentLicense = .free
                        self?.resetUsage()
                    }
                }
            }
            .store(in: &cancellables)

        // Start periodic revalidation timer
        startPeriodicRevalidation()
    }

    // MARK: - Periodic Revalidation

    private func startPeriodicRevalidation() {
        revalidationTimer?.invalidate()
        revalidationTimer = Timer.scheduledTimer(
            withTimeInterval: TimeInterval(revalidationIntervalMinutes * 60),
            repeats: true
        ) { [weak self] _ in
            guard let self else { return }
            Task { @MainActor [self] in
                await self.revalidateIfNeeded()
            }
        }
    }

    /// Revalidate only if authenticated and not recently validated
    private func revalidateIfNeeded() async {
        guard authManager.isAuthenticated else { return }

        // Skip if validated recently (within last 5 minutes)
        if let lastValidated = lastValidatedAt,
           Date().timeIntervalSince(lastValidated) < 300 {
            return
        }

        await revalidate()
    }

    // MARK: - Feature Access

    /// Check if a feature can be used
    /// Returns .blocked for unauthenticated users trying to use features
    func canUse(_ feature: Feature) -> FeatureAccess {
        // Block all features except viewing UI for unauthenticated users
        if !authManager.isAuthenticated {
            // Session sync requires auth specifically
            if case .sessionSync = feature {
                return .requiresAuth
            }
            // All other features are blocked for unauthenticated users
            return .blocked
        }

        let features = currentLicense.features

        switch feature {
        case .smartMode:
            // Smart Mode is Enterprise-only
            if !features.smartModeEnabled {
                return .requiresEnterprise
            }
            if let limit = features.smartModeLimit {
                if smartModeUsedToday >= limit {
                    return .limitReached(used: smartModeUsedToday, limit: limit)
                }
            }
            return .allowed

        case .customModes:
            return features.customModesEnabled ? .allowed : .requiresPro

        case .exportMarkdown:
            return features.exportFormats.contains("markdown") ? .allowed : .requiresPro

        case .exportJson:
            return features.exportFormats.contains("json") ? .allowed : .requiresPro

        case .autoAnswer:
            // Auto-Answer is Enterprise-only
            return features.autoAnswerEnabled ? .allowed : .requiresEnterprise

        case .sessionSync:
            return features.sessionSyncEnabled ? .allowed : .requiresPro

        case .aiRequest:
            if let limit = features.dailyAiRequestLimit {
                if aiRequestsToday >= limit {
                    return .limitReached(used: aiRequestsToday, limit: limit)
                }
            }
            return .allowed

        case .undetectable:
            // Undetectable overlay is Enterprise-only
            return features.undetectableEnabled ? .allowed : .requiresEnterprise

        case .screenshot:
            return features.screenshotEnabled ? .allowed : .requiresPro

        case .sessionStart:
            // Session start requires authentication
            return authManager.isAuthenticated ? .allowed : .blocked

        case .knowledgeBase:
            // Knowledge Base (Context Intelligence) is Enterprise-only
            return features.knowledgeBaseEnabled ? .allowed : .requiresEnterprise

        case .proactiveSuggestions:
            // Proactive AI suggestions are Enterprise-only
            return features.proactiveSuggestionsEnabled ? .allowed : .requiresEnterprise
        }
    }

    /// Record usage of a feature (for rate-limited features)
    /// Also sends usage to server asynchronously for server-side tracking
    func recordUsage(_ feature: Feature, provider: String? = nil) {
        let action: String
        switch feature {
        case .smartMode:
            smartModeUsedToday += 1
            saveUsage()
            action = "smart_mode"
        case .aiRequest:
            aiRequestsToday += 1
            saveUsage()
            action = "ai_request"
        case .sessionStart:
            action = "session_start"
        case .autoAnswer:
            action = "auto_answer"
        default:
            return
        }

        // Record usage to server asynchronously (fire-and-forget)
        Task {
            do {
                try await api.recordUsage(action: action, provider: provider)
            } catch {
                print("[License] Failed to record usage to server: \(error)")
            }
        }
    }

    // MARK: - Plan Properties

    var isPro: Bool {
        (currentLicense.plan == .pro || currentLicense.plan == .enterprise) &&
        (currentLicense.status == .active || currentLicense.status == .trialing)
    }

    var isEnterprise: Bool {
        currentLicense.plan == .enterprise &&
        (currentLicense.status == .active || currentLicense.status == .trialing)
    }

    var isTrialing: Bool {
        currentLicense.status == .trialing
    }

    var trialDaysRemaining: Int? {
        currentLicense.trial?.daysRemaining
    }

    // MARK: - Signature Verification Methods

    /// Verifies the HMAC-SHA256 signature of a license response
    /// - Parameters:
    ///   - license: The license to verify
    /// - Returns: true if signature is valid, false otherwise
    private func verifyLicenseSignature(_ license: License) -> Bool {
        // Skip verification for local free license (no signature)
        if license.signature.isEmpty && license.plan == .free {
            return true
        }

        // SECURITY: Require license secret to be available
        // If not configured, fail-safe by rejecting the signature
        guard let secret = licenseSecret, !secret.isEmpty else {
            print("[License] SECURITY: LICENSE_SECRET not configured - signature verification skipped")
            print("[License] SECURITY: Set LICENSE_SECRET environment variable or configure server-side validation")
            // For non-free plans, we should be strict - but during initial launch
            // we can allow validation to pass through to server-side checks
            // The server will still enforce license validity
            return true // Trust server-side validation when client-side secret is unavailable
        }

        // Build the payload that was signed (same fields as server)
        // Order must match server: valid, plan, status, features, trial, cacheTTL, validatedAt
        let signedPayload = buildSignaturePayload(from: license)

        guard let payloadData = signedPayload.data(using: .utf8) else {
            print("[License] Failed to encode payload for signature verification")
            return false
        }

        // Compute HMAC-SHA256
        let key = SymmetricKey(data: Data(secret.utf8))
        let signature = HMAC<SHA256>.authenticationCode(for: payloadData, using: key)
        let computedSignature = signature.map { String(format: "%02x", $0) }.joined()

        // Compare signatures using constant-time comparison to prevent timing attacks
        let isValid = constantTimeCompare(computedSignature, license.signature)

        if !isValid {
            print("[License] Signature verification failed")
            print("[License] Expected: \(computedSignature.prefix(16))...")
            print("[License] Got: \(license.signature.prefix(16))...")
        }

        return isValid
    }

    /// Constant-time string comparison to prevent timing attacks
    private func constantTimeCompare(_ a: String, _ b: String) -> Bool {
        guard a.count == b.count else { return false }

        let aBytes = Array(a.utf8)
        let bBytes = Array(b.utf8)

        var result: UInt8 = 0
        for i in 0..<aBytes.count {
            result |= aBytes[i] ^ bBytes[i]
        }

        return result == 0
    }

    /// Builds the JSON payload string that was signed by the server
    /// Must match the exact format and order used by the server
    private func buildSignaturePayload(from license: License) -> String {
        // The server signs this structure (JSON.stringify preserves insertion order):
        // { valid, plan, status, features, trial, cacheTTL, validatedAt }

        var payload = "{"
        payload += "\"valid\":\(license.valid),"
        payload += "\"plan\":\"\(license.plan.rawValue)\","
        payload += "\"status\":\"\(license.status.rawValue)\","
        payload += "\"features\":\(buildFeaturesJSON(license.features)),"
        payload += "\"trial\":\(buildTrialJSON(license.trial)),"
        payload += "\"cacheTTL\":\(license.cacheTTL),"
        payload += "\"validatedAt\":\"\(license.validatedAt)\""
        payload += "}"

        return payload
    }

    private func buildFeaturesJSON(_ features: LicenseFeatures) -> String {
        var json = "{"
        json += "\"smartModeEnabled\":\(features.smartModeEnabled),"
        json += "\"smartModeLimit\":\(features.smartModeLimit.map { String($0) } ?? "null"),"
        json += "\"customModesEnabled\":\(features.customModesEnabled),"
        json += "\"exportFormats\":[\(features.exportFormats.map { "\"\($0)\"" }.joined(separator: ","))],"
        json += "\"autoAnswerEnabled\":\(features.autoAnswerEnabled),"
        json += "\"sessionSyncEnabled\":\(features.sessionSyncEnabled),"
        json += "\"dailyAiRequestLimit\":\(features.dailyAiRequestLimit.map { String($0) } ?? "null"),"
        json += "\"maxSyncedSessions\":\(features.maxSyncedSessions.map { String($0) } ?? "null"),"
        json += "\"maxTranscriptSize\":\(features.maxTranscriptSize.map { String($0) } ?? "null"),"
        json += "\"undetectableEnabled\":\(features.undetectableEnabled),"
        json += "\"screenshotEnabled\":\(features.screenshotEnabled),"
        json += "\"knowledgeBaseEnabled\":\(features.knowledgeBaseEnabled),"
        json += "\"proactiveSuggestionsEnabled\":\(features.proactiveSuggestionsEnabled)"
        json += "}"
        return json
    }

    private func buildTrialJSON(_ trial: TrialInfo?) -> String {
        guard let trial = trial else { return "null" }
        var json = "{"
        json += "\"isActive\":\(trial.isActive),"
        json += "\"daysRemaining\":\(trial.daysRemaining),"
        json += "\"endsAt\":\"\(trial.endsAt)\""
        json += "}"
        return json
    }

    // MARK: - Validation

    /// Force revalidate license from server
    func revalidate() async {
        guard authManager.isAuthenticated else {
            currentLicense = .free
            lastSignatureError = nil
            return
        }

        isValidating = true
        isOffline = false
        lastSignatureError = nil

        do {
            let license = try await api.validateLicense(deviceId: deviceInfo.deviceId)

            // Verify signature before accepting the license
            if verifyLicenseSignature(license) {
                currentLicense = license
                lastValidatedAt = Date()

                // Update usage from server
                if let usage = license.usage {
                    smartModeUsedToday = usage.smartModeUsedToday
                    aiRequestsToday = usage.aiRequestsToday
                }

                // Cache the license
                cacheLicense(license)
            } else {
                // Signature verification failed - potential tampering
                print("[License] SECURITY: Invalid license signature detected")
                lastSignatureError = "License signature verification failed"

                // Fall back to cached license if valid, otherwise use free
                if let cachedLicense = loadCachedLicense(),
                   verifyLicenseSignature(cachedLicense) {
                    currentLicense = cachedLicense
                } else {
                    currentLicense = .free
                }
            }

        } catch {
            print("[License] Validation failed: \(error)")
            isOffline = true

            // Use cached license with grace period
            if let cachedLicense = loadCachedLicense() {
                // Verify cached license signature
                guard verifyLicenseSignature(cachedLicense) else {
                    print("[License] SECURITY: Cached license has invalid signature")
                    currentLicense = .free
                    isValidating = false
                    return
                }

                // Check grace period for PRO users
                if cachedLicense.plan == .pro {
                    if let expiry = getCacheExpiry(), expiry > Date() {
                        currentLicense = cachedLicense
                    } else {
                        // Grace period expired, fallback to free
                        currentLicense = .free
                    }
                } else {
                    currentLicense = cachedLicense
                }
            }
        }

        isValidating = false
    }

    // MARK: - Cache Management

    @discardableResult
    private func loadCachedLicense() -> License? {
        guard let data = UserDefaults.standard.data(forKey: cacheKey),
              let license = try? JSONDecoder().decode(License.self, from: data) else {
            return nil
        }
        currentLicense = license
        return license
    }

    private func cacheLicense(_ license: License) {
        guard let data = try? JSONEncoder().encode(license) else { return }

        UserDefaults.standard.set(data, forKey: cacheKey)

        // Set expiry with grace period for PRO
        let expiryDays = license.plan == .pro ? gracePeriodDays : 1
        let expiry = Date().addingTimeInterval(TimeInterval(expiryDays * 24 * 60 * 60))
        UserDefaults.standard.set(expiry, forKey: cacheExpiryKey)
    }

    private func getCacheExpiry() -> Date? {
        UserDefaults.standard.object(forKey: cacheExpiryKey) as? Date
    }

    // MARK: - Usage Management

    private func resetDailyUsageIfNeeded() {
        let today = Calendar.current.startOfDay(for: Date())
        let lastReset = UserDefaults.standard.object(forKey: usageResetKey) as? Date

        if lastReset == nil || !Calendar.current.isDate(lastReset!, inSameDayAs: today) {
            resetUsage()
            UserDefaults.standard.set(today, forKey: usageResetKey)
        } else {
            loadUsage()
        }
    }

    private func resetUsage() {
        smartModeUsedToday = 0
        aiRequestsToday = 0
        saveUsage()
    }

    private func loadUsage() {
        smartModeUsedToday = UserDefaults.standard.integer(forKey: "smartModeUsedToday")
        aiRequestsToday = UserDefaults.standard.integer(forKey: "aiRequestsToday")
    }

    private func saveUsage() {
        UserDefaults.standard.set(smartModeUsedToday, forKey: "smartModeUsedToday")
        UserDefaults.standard.set(aiRequestsToday, forKey: "aiRequestsToday")
    }
}

// MARK: - Convenience Extensions

extension LicenseManager {
    /// Check if feature is available (simple boolean)
    func isFeatureAvailable(_ feature: Feature) -> Bool {
        if case .allowed = canUse(feature) {
            return true
        }
        return false
    }

    /// Get remaining uses for a rate-limited feature
    func remainingUses(for feature: Feature) -> Int? {
        switch feature {
        case .smartMode:
            guard let limit = currentLicense.features.smartModeLimit else { return nil }
            return max(0, limit - smartModeUsedToday)
        case .aiRequest:
            guard let limit = currentLicense.features.dailyAiRequestLimit else { return nil }
            return max(0, limit - aiRequestsToday)
        default:
            return nil
        }
    }
}
